map.js

```diff
@@ -467,66 +415,108 @@ function initListeners(){
         lastMouseLoc = new Point(event.offsetX, event.offsetY);
         isDown = false;
         //yay we get to interpret clicks on our own! /s
-        if(Math.abs(downLoc.x - event.offsetX) < CLICK_LIMIT_PIXELS &&
-            Math.abs(downLoc.y - event.offsetY) < CLICK_LIMIT_PIXELS &&
-            Date.now() - clickStart < CLICK_LIMIT_DOWN_MS){
+        if(Math.abs(mouseDownLoc.x - event.offsetX) < CLICK_LIMIT_PIXELS &&
+            Math.abs(mouseDownLoc.y - event.offsetY) < CLICK_LIMIT_PIXELS &&
+            Date.now() - clickStartTime < CLICK_LIMIT_DOWN_MS){
                 onMouseClick(lastMouseLoc);
         }
         //TODO: handle double clicks
     });
     viewport.onmouseout = function(){isDown = false;};
     viewport.onwheel = function(e){    
+        
         e.preventDefault();
-        if(e.deltaY > 0) zoomLevel += 0.2;
-        else zoomLevel += -0.2;
+        const zoomPoint = new Point(e.offsetX, e.offsetY);
+        if(e.deltaY > 0){ 
+        updateZoom(0.2, zoomPoint);
+        }
+        else {
+            updateZoom(-0.2, zoomPoint);
+        }
         
-        //TODO: make it so that it zooms into middle of screen rather than top left corner?
-
-        //clamp zoom
-        if(zoomLevel > maxZoom) zoomLevel = maxZoom;
-        if(zoomLevel < minZoom) zoomLevel = minZoom;
-        moveMap();
         drawFrame();
     };
 }
 
+function updateZoom(deltaZ, zoomPoint){
+    const ICON_NATIVE_HEIGHT = 20;
+    let oldZoom = zoomLevel;
+    zoomLevel += deltaZ;
+    //clamp zoom
+    if(zoomLevel > maxZoom) zoomLevel = maxZoom;
+    if(zoomLevel < minZoom) zoomLevel = minZoom;
+
+    if(oldZoom == zoomLevel){
+        return;
+    }
+
+    //adjust icon size
+    var m_iconH = ICON_NATIVE_HEIGHT / zoomLevel;
+    if(zoomLevel > 1.75)m_iconH = ICON_NATIVE_HEIGHT / zoomLevel * 2;
+    else if(zoomLevel > 1.5)m_iconH = ICON_NATIVE_HEIGHT / zoomLevel * 1.5;
+    else if(zoomLevel > 1.25)m_iconH = ICON_NATIVE_HEIGHT / zoomLevel * 1.25;
+    _iconH = m_iconH;
+
+    //make map zoom in to zoomPoint.
+    //1: calculate current zoomPoint in map coords
+    //2. calculate where that point is on the new map
+    //3. calculate where the corner needs to be to set that point as the center
+    let oldCenterMapCoord = screenSpaceToMapSpace(zoomPoint);
+    let newCenterMapCoord = oldCenterMapCoord.multiply(oldZoom/zoomLevel);
+    let newCornerMapCoord = newCenterMapCoord.subtract(zoomPoint);
+
+    //moveMap takes a delta, so we subtract new from old. 
+    moveMap(screenOriginInMapCoords.subtract(newCornerMapCoord));
+}
+
 //converts worldspace cords into map coords.
 //this is a pixel measurement from upper left of map image.
-function worldSpaceToMapSpace(x = 0, y = 0){
+function worldSpaceToMapSpace(point){
     //first, we convert world space into map space.
     var MapW = img_Map.width;
     var MapH = img_Map.height;
     const worldW = 480000;
     const worldH = 400000;
-    const ICON_NATIVE_HEIGHT = 20;
```

Comment:
Moved icon height changing to zooming function, as thats the only time it changes.